storage/tianmu/handler/ha_tianmu.cpp

```diff
@@ -1780,41 +1780,59 @@ void ha_tianmu::key_convert(const uchar *key, uint key_len, std::vector<uint> co
     std::unique_ptr<char[]> buf(new char[length]);
     char *ptr = buf.get();
 
+    // TODO(gry): why truncate when v out of range? Is here need to keep consistent with write data?
+    auto unsigned_flag = f->flags & UNSIGNED_FLAG;
     switch (f->type()) {
       case MYSQL_TYPE_TINY: {
         int64_t v = f->val_int();
-        if (v > TIANMU_TINYINT_MAX)
-          v = TIANMU_TINYINT_MAX;
-        else if (v < TIANMU_TINYINT_MIN)
-          v = TIANMU_TINYINT_MIN;
-        *(int64_t *)ptr = v;
+        if (unsigned_flag) {
+          v = (v > UINT_MAX8) ? UINT_MAX8 : v;
+        } else {
+          if (v > TIANMU_TINYINT_MAX)
```

Comment:
> How about if v is at the boundary of tianmu_tiny; [TIANMU_TINYINT_MAX, TIANM_TINYINT_MIN]。If the boundery uses for NULL flag. Pls give more comments on this, othwise, you will forget "why".

You are right, I'll add some comments to interprete if `NULL` has impact on boundary.

Reply:
> if (xx) a else if(yy) b. can be refined as (xx) ? a : (yyy) ? b : c;

Not a good idea, I had thought about it before.
The best way is do refactor, add new data types named `utinyint`, `usmallint`, `umediumint`, `uint`, `uint64`, just like MariaDB ColumnStore and other dbs does.

Reply:
> > if (xx) a else if(yy) b. can be refined as (xx) ? a : (yyy) ? b : c;
> 
> Not a good idea, I had thought about it before. The best way is do refactor, add new data types named `utinyint`, `usmallint`, `umediumint`, `uint`, `uint64`, just like MariaDB ColumnStore and other dbs does. Only in this way can make the code be more elegant.

OK

Reply:
> > > if (xx) a else if(yy) b. can be refined as (xx) ? a : (yyy) ? b : c;
> > 
> > 
> > Not a good idea, I had thought about it before. The best way is do refactor, add new data types named `utinyint`, `usmallint`, `umediumint`, `uint`, `uint64`, just like MariaDB ColumnStore and other dbs does. Only in this way can make the code be more elegant.
> 
> OK

I think, refactor it will be a good choice, if we can finish these jobs in a few days.